src/python/pants/jvm/goals/lockfile.py

```diff
@@ -16,63 +20,212 @@
 from pants.jvm.resolve.jvm_tool import GatherJvmCoordinatesRequest, JvmToolBase
-from pants.jvm.resolve.key import CoursierResolveKey
 from pants.jvm.resolve.lockfile_metadata import JVMLockfileMetadata
 from pants.jvm.subsystems import JvmSubsystem
-from pants.jvm.target_types import JvmArtifactCompatibleResolvesField
+from pants.jvm.target_types import JvmArtifactCompatibleResolvesField, JvmArtifactJarSourceField
+from pants.jvm.util_rules import ExtractFileDigest
 from pants.util.logging import LogLevel
 from pants.util.meta import frozen_after_init
 from pants.util.ordered_set import FrozenOrderedSet
+from pants.util.strutil import pluralize
+
+# -----------------------------------------------------------------------
+# Tool lockfiles
+# -----------------------------------------------------------------------
 
 
 @dataclass(frozen=True)
 class GenerateJvmLockfile(GenerateLockfile):
     artifacts: ArtifactRequirements
 
 
-@frozen_after_init
-@dataclass(unsafe_hash=True)
-class GenerateJvmLockfileFromTool:
-    artifact_inputs: FrozenOrderedSet[str]
-    options_scope: str
-    lockfile_dest: str
+@rule
+async def prepare_coursier_resolve_info(
+    artifact_requirements: ArtifactRequirements,
+) -> CoursierResolveInfo:
+    # Transform requirements that correspond to local JAR files into coordinates with `file:/`
+    # URLs, and put the files in the place specified by the URLs.
+    no_jars: List[ArtifactRequirement] = []
+    jars: List[Tuple[ArtifactRequirement, JvmArtifactJarSourceField]] = []
 
-    def __init__(self, tool: JvmToolBase) -> None:
-        # Note that `JvmToolBase` is not hashable, so we extract the relevant information eagerly.
-        self.artifact_inputs = FrozenOrderedSet(tool.artifact_inputs)
-        self.options_scope = tool.options_scope
-        self.lockfile_dest = tool.lockfile
+    for req in artifact_requirements:
+        jar = req.jar
+        if not jar:
+            no_jars.append(req)
+        else:
+            jars.append((req, jar))
 
+    jar_files = await Get(SourceFiles, SourceFilesRequest(i[1] for i in jars))
+    jar_file_paths = jar_files.snapshot.files
 
-@rule
-async def setup_lockfile_request_from_tool(
-    request: GenerateJvmLockfileFromTool,
-) -> GenerateJvmLockfile:
-    artifacts = await Get(
-        ArtifactRequirements,
-        GatherJvmCoordinatesRequest(
-            request.artifact_inputs,
-            f"[{request.options_scope}].artifacts",
+    resolvable_jar_requirements = [
+        dataclasses.replace(
+            req, jar=None, url=f"file:{Coursier.working_directory_placeholder}/{path}"
+        )
+        for req, path in zip((i[0] for i in jars), jar_file_paths)
+    ]
+
+    to_resolve = chain(no_jars, resolvable_jar_requirements)
+
+    return CoursierResolveInfo(
+        coord_arg_strings=frozenset(req.to_coord_arg_str() for req in to_resolve),
+        digest=jar_files.snapshot.digest,
+    )
+
+
+@rule(level=LogLevel.DEBUG)
+async def coursier_resolve_lockfile(
```

Comment:
At this point, I might rename this to `coursier_resolve_lockfile_from_requirements`, just to make it really clear what's calling this.